### ISSUE TITLE
Virtual keyboard missing keys and not showing up in any emulator

### ISSUE THREAD

**Issue by rsn8887** [libretro/RetroArch#6209]
## Description

When binding any key to "on-screen keyboard" in input->hotkey binds, the virtual keyboard only shows up in the gui, but not in the emulators. The virtual keyboard does not show all the keys that are present on a physical keyboard. For example, keypad keys, control, alt etc. are all missing.

### Expected behavior

The virtual keyboard should not only appear in the gui, but also show up inside the emulators such as Dosbox and Vice when I press the hotkey. Furthermore, it should allow me to select and press keys using a game controller. Those should then be send to the emulator core. Basically the virtual keyboard should act as if a real physical keyboard was connected and those keys were pressed on it at any place within Retroarch (gui, emulators etc.)

### Actual behavior

see description

### Steps to reproduce the bug

1. bind hotkey to on-screen keyboard toggle
2. press hotkey inside gui, virtual keyboard appears but is missing keys like alt, ctrl etc.
3. press hotkey inside emulator, virtual keyboard does not appear

### Bisect Results

The virtual keyboard basically never worked for me in any useful way apart from limited functionality inside the gui itself.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 1.7.0 nightly

### Environment information

- OS: RPi, Vita, PC (tested on those three)

**Comment by Ferk** [comment]
The virtual keyboard should probably be a special case of overlay that can be navigated with the gamepad and can have a hotkey to bring it up, even while in the GUI.
It makes sense for it to be customizable just like other overlays, so people in non-english languages can add set a custom overlay as osk to add their own symbols.

**Comment by i30817** [comment]
Bumping this, because it would be really rally useful in some cores for phone devices. I mean, some cores require you press buttons but not necessarily in a frenetic way that would require button mapping.

For instance imagine a Adventure Game Studio core, or even the scummvm core. It would be very good if the virtual keyboard could seamlessly plugin as the device without any focus bugs, and would easily reduce port complexity.

**Comment by Zlika** [comment]
llibretro-uae and libretro-vice use a custom "VKDB" virtual keyboard feature (cf. https://github.com/libretro/vice-libretro/commit/d74caf5c7789ee0c8434be710e2f71af606543a2). I think it would be interesting to add this feature to libretro-common to make it easily available for all cores.

**Comment by rsn8887** [comment]
Yes, thanks to the heroic effort of @Sonninnos who worked tirelessly on Vice and PUAE, we now have really good virtual keyboards in those cores.

My issue here was created before all that work was done. I was hoping for a universal virtual keyboard provided by the frontend that could work in any core, even those that don't have their own vkbd, for example dosbox-svn.

**Comment by Spawk-HS** [comment]
Any progress on a dosbox VKBD? Almost every one of my games require typing at least at some point... enter your name, save game name! Being able to load them is great but the core is not actually "usable" at this point.

**Comment by rsn8887** [comment]
Well, you can use a BT keyboard on Vita and a USB keyboard on Switch for Dosbox SVN, but yeah, not very usable on Mobile platforms.

**Comment by Ultrahead** [comment]
Well, this is an old thread but if you need and OSK, maybe you can fork and try to integrate [this virtual keyboard](https://github.com/Ultrahead/virtualkeyboard), based on SDL2, tailoring it for your needs.

Edit: feel free to modify the look and feel, and even the functionality, for your needs.

**Comment by zoltanvb** [comment]
Old thread indeed. Nowadays, the virtual keyboard of the menu system is a separate entity, and will not appear for cores. However, it is possible to set up a keyboard overlay, that should work on any core. A hotkey (button) can be assigned, it is the last entry among hotkeys, but it requires 2 things:
- overlays enabled
- keyboard overlay set - it is a separate, second overlay, the menu entry appears only under certain conditions... this is probably something that can be improved.

**Comment by Ultrahead** [comment]
Improving this could be indeed great for cores like the MSX ones that do not come with OSK.

**Comment by Ultrahead** [comment]
Once the keyboard overlay appears on the screen, is there a way to use the game pad to select and press a key in it?

**Comment by zoltanvb** [comment]
Oh right, it is not possible to use the overlay keyboard if only a gamepad is present. Hmm.

**Comment by Ultrahead** [comment]
Any plans to add supporting the use of the keyboard overlay when only a gamepad is present?

**Comment by Ultrahead** [comment]
Anyone?
